datafusion/sql/src/unparser/rewrite.rs

```diff
@@ -257,6 +257,37 @@ pub(super) fn subquery_alias_inner_query_and_columns(
     (outer_projections.input.as_ref(), columns)
 }
 
+/// Injects column aliases into the projection of a logical plan by wrapping `Expr::Column` expressions
+/// with `Expr::Alias` using the provided list of aliases. Non-column expressions are left unchanged.
+///
+/// Example:
+/// - `SELECT col1, col2 FROM table` with aliases `["alias_1", "some_alias_2"]` will be transformed to
+/// - `SELECT col1 AS alias_1, col2 AS some_alias_2 FROM table`
+pub(super) fn inject_column_aliases(
+    projection: &datafusion_expr::Projection,
+    aliases: &Vec<Ident>,
+) -> LogicalPlan {
+    let mut updated_projection = projection.clone();
+
+    let new_exprs = projection
+        .expr
+        .iter()
+        .zip(aliases)
+        .map(|(expr, col_alias)| match expr {
+            Expr::Column(col) => Expr::Alias(Alias {
+                expr: Box::new(expr.clone()),
+                relation: col.relation.clone(),
+                name: col_alias.to_string(),
+            }),
+            _ => expr.clone(),
+        })
+        .collect::<Vec<_>>();
+
+    updated_projection.expr.clone_from(&new_exprs);
```

Comment:
We can avoid a few unnecessary clones by accepting an `impl IntoIterator<Ident>` for `aliases` (since the caller doesn't actually need them anymore). Then the calling code becomes:

```rust
let rewritten_plan = inject_column_aliases(inner_p, columns);
columns = vec![];
```

And this can be:

```rust
pub(super) fn inject_column_aliases(
    projection: &datafusion_expr::Projection,
    aliases: impl IntoIterator<Item = Ident>,
) -> LogicalPlan {
    let mut updated_projection = projection.clone();

    let new_exprs = updated_projection
        .expr
        .into_iter()
        .zip(aliases)
        .map(|(expr, col_alias)| match expr {
            Expr::Column(col) => {
                let relation = col.relation.clone();
                Expr::Alias(Alias {
                    expr: Box::new(Expr::Column(col)),
                    relation,
                    name: col_alias.value,
                })
            }
            _ => expr,
        })
        .collect::<Vec<_>>();

    updated_projection.expr = new_exprs;

    LogicalPlan::Projection(updated_projection)
}
```

Reply:
Cloning `TableReference` is cheap, so `col.relation.clone()` is lightweight. The only clone we need to do now is `projection.clone()` which I don't think there is a way around.

Reply:
Updated as suggested - thank you 👍